### PR TITLE
Handle exceeded child records during EVM execution

### LINKED ISSUE #4370: Update the traceability records accordingly for multiple calls to the precompiled contracts
Add logic to handle `ResourceLimitException`s and externalize appropriate sidecar records.

### PR DESCRIPTION
**Pull request by dimitar-dinev**
**Description**:

* Adds `ResourceLimitExeception` handling [around EVM execution processing ](https://github.com/hashgraph/hedera-services/pull/4382/files#diff-bfc28a0f15a0a1e48d14b402e5b25fbd28e8eb796b65a746312f5bb1ae76cfbcR139)the same way [we do on `updater.commit()`](https://github.com/hashgraph/hedera-services/pull/4382/files#diff-bfc28a0f15a0a1e48d14b402e5b25fbd28e8eb796b65a746312f5bb1ae76cfbcR206)

**Related issue(s)**:

Fixes #4370

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 1]
```diff
@@ -336,4 +337,48 @@ private ChargingResult chargeForGas(
...
+                feesOnlyUpdater);
+        sendToCoinbase(coinbase, gasLimit, gasPrice, feesOnlyUpdater);
+        // We can't go through the top-level commit() because that would
+        // re-try to commit the storage changes
+        feesOnlyUpdater.trackingAccounts().commit();
+    }
+
+    private TransactionProcessingResult createResourceLimitExceededResult(
+            final long gasPrice, final long gasLimit, final ResourceLimitException e) {
+        return TransactionProcessingResult.failed(
```
Comment (david-bakin-sl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:375]: (it's a shame `TransactionProcessingResult` doesn't have a builder ... all these empty parameters and the reader either has to _know_ what they are or go look it up)

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 2]
```diff
@@ -124,16 +124,32 @@ protected TransactionProcessingResult execute(
...
+                    sender,
+                    receiver,
+                    gasPrice,
+                    gasLimit,
+                    value,
+                    payload,
+                    contractCreation,
+                    isStatic,
+                    mirrorReceiver);
+        } catch (ResourceLimitException e) {
```
Comment (david-bakin-sl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:139]: line 191 has a `final` keyword here for the exception?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 3]
```diff
@@ -124,16 +124,32 @@ protected TransactionProcessingResult execute(
...
+            handleResourceLimitExceeded(
```
Comment (david-bakin-sl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:140]: after handling a 'resource limit exceeded' you're always going to throw a proper exception for it - so make it one step?  (`handleResourceLimitExceeded` to return the exception to throw?)

File: test-clients/src/main/resource/contract/contracts/ERC20Contract/ERC20Contract.sol [hunk 4]
```diff
@@ -9,6 +9,11 @@ contract ERC20Contract {
         IERC20Metadata(token).name();
     }
 
+    function nameNTimes(address token, uint times) public view {
+        for (uint i = 0; i < times; i++) {
+            IERC20Metadata(token).name();
+        }
+    }
     function symbol(address token) public view {
```
Comment (david-bakin-sl) [inline test-clients/src/main/resource/contract/contracts/ERC20Contract/ERC20Contract.sol:17]: (nit: add blank line separator)

**Review by david-bakin-sl — COMMENTED** [review]
(superficial review of tiny local stuff as I don't know the domain - thus no "approve" only since I'm not qualified)

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java (hunk 3 above)
Comment (dimitar-dinev) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java]: By 'throwing a proper exception` you have in mind returning the `TransactionProcessingResult`, i.e. combining the 2 functions, right? I wanted to somewhat separate the actions of 'teardown' after processing and the return of the specific processing result, but the suggested approach (if I'm understanding correctly) sounds good to me also

File: test-clients/src/main/resource/contract/contracts/ERC20Contract/ERC20Contract.sol (hunk 4 above)
Comment (dimitar-dinev) [reply test-clients/src/main/resource/contract/contracts/ERC20Contract/ERC20Contract.sol:18]: Added

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ERCPrecompileSuite.java [hunk 5]
```diff
@@ -184,7 +189,7 @@ public boolean canRunConcurrent() {
 
     @Override
     public List<HapiApiSpec> getSpecsInSuite() {
-        return allOf(ERC_20(), ERC_721());
```
[the remark this answers is not in the document]
Comment (dimitar-dinev) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ERCPrecompileSuite.java]: Oh, good catch --- I had removed it for ease of local testing and then forgot to bring it back

**Review by tinker-michaelj — APPROVED** [review]
Nice. 😎

**Review by stoqnkpL — APPROVED** [review]

**Pull request merged**